pkg/ingester/circuitbreaker_test.go

```diff
@@ -0,0 +1,816 @@
+				# TYPE cortex_ingester_circuit_breaker_results_total counter
+				cortex_ingester_circuit_breaker_results_total{result="success"} 0
+				cortex_ingester_circuit_breaker_results_total{result="error"} 0
+				cortex_ingester_circuit_breaker_results_total{result="circuit_breaker_open"} 0
+			`,
+		},
+		"with a permit acquired, pushRequestDuration higher than PushTimeout and an input error different from context deadline exceeded, finishPushRequest gives context deadline exceeded error": {
+			pushRequestDuration:          3 * time.Second,
+			acquiredCircuitBreakerPermit: true,
+			err:                          newInstanceLimitReachedError("error"),
+			expectedMetrics: `
+				# HELP cortex_ingester_circuit_breaker_results_total Results of executing requests via the circuit breaker.
+				# TYPE cortex_ingester_circuit_breaker_results_total counter
+				cortex_ingester_circuit_breaker_results_total{result="success"} 0
+				cortex_ingester_circuit_breaker_results_total{result="error"} 1
+				cortex_ingester_circuit_breaker_results_total{result="circuit_breaker_open"} 0
+			`,
+		},
+		"with a permit not acquired, pushRequestDuration higher than PushTimeout and an input error different from context deadline exceeded, finishPushRequest does nothing": {
+			pushRequestDuration:          3 * time.Second,
+			acquiredCircuitBreakerPermit: false,
+			err:                          newInstanceLimitReachedError("error"),
+			expectedMetrics: `
+				# HELP cortex_ingester_circuit_breaker_results_total Results of executing requests via the circuit breaker.
+				# TYPE cortex_ingester_circuit_breaker_results_total counter
+				cortex_ingester_circuit_breaker_results_total{result="success"} 0
+				cortex_ingester_circuit_breaker_results_total{result="error"} 0
+				cortex_ingester_circuit_breaker_results_total{result="circuit_breaker_open"} 0
+			`,
+		},
+	}
+	for testName, testCase := range testCases {
+		reg := prometheus.NewPedanticRegistry()
+		cfg := defaultIngesterTestConfig(t)
+		cfg.CircuitBreakerConfig = CircuitBreakerConfig{
+			Enabled:     true,
+			PushTimeout: 2 * time.Second,
+		}
+
+		i, err := prepareIngesterWithBlocksStorage(t, cfg, nil, reg)
+		require.NoError(t, err)
+
+		require.NoError(t, services.StartAndAwaitRunning(context.Background(), i))
+		defer services.StopAndAwaitTerminated(context.Background(), i) //nolint:errcheck
+
+		// Wait until the ingester is healthy
+		test.Poll(t, 100*time.Millisecond, 1, func() interface{} {
+			return i.lifecycler.HealthyInstancesCount()
+		})
+
+		ctx := user.InjectOrgID(context.Background(), "test")
+
+		st := &pushRequestState{
+			requestDuration:              testCase.pushRequestDuration,
+			acquiredCircuitBreakerPermit: testCase.acquiredCircuitBreakerPermit,
+			pushErr:                      testCase.err,
+		}
+		ctx = context.WithValue(ctx, pushReqCtxKey, st)
+
+		t.Run(testName, func(t *testing.T) {
+			i.FinishPushRequest(ctx)
```

Comment:
I'd suggest to put entire test setup into `t.Run` function, so that any possible errors above are reported in correct sub-test.

Reply:
This wasn't intentional. Thank you for finding this